flow/layers/physical_model_layer.cc

```diff
@@ -0,0 +1,53 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/flow/layers/physical_model_layer.h"
+
+#include "third_party/skia/include/utils/SkShadowUtils.h"
+
+namespace flow {
+
+PhysicalModelLayer::PhysicalModelLayer()
+    : rrect_(SkRRect::MakeEmpty()) {}
+
+PhysicalModelLayer::~PhysicalModelLayer() {}
+
+void PhysicalModelLayer::Preroll(PrerollContext* context, const SkMatrix& matrix) {
+  PrerollChildren(context, matrix);
+  if (!context->child_paint_bounds.intersect(rrect_.getBounds()))
+    context->child_paint_bounds.setEmpty();
+
+  // Add some margin to the paint bounds to leave space for the shadow.
+  SkRect bounds(context->child_paint_bounds);
+  bounds.outset(50.0, 50.0);
+  set_paint_bounds(bounds);
+}
+
+void PhysicalModelLayer::Paint(PaintContext& context) {
+  TRACE_EVENT0("flutter", "PhysicalModelLayer::Paint");
+  FTL_DCHECK(!needs_system_composite());
+
+  SkPath path;
+  path.addRRect(rrect_);
+
+  if (elevation_ != 0) {
+    SkShadowUtils::DrawShadow(&context.canvas, path,
+                              elevation_ * 4,
+                              SkPoint3::Make(0.0f, -700.0f, 2800.0f),
+                              2800.0f,
+                              0.25f, 0.25f,
+                              SK_ColorBLACK);
+  }
+
+  SkPaint paint;
+  paint.setColor(SkColorSetA(color_, 0xFF));
+  context.canvas.drawPath(path, paint);
+
+  SkAutoCanvasRestore save(&context.canvas, false);
+  context.canvas.saveLayer(&paint_bounds(), nullptr);
```

Comment:
We should check whether the rrect is actually a rect (i.e., whether the radius is zero) and use save() rather than saveLayer() in that case.

Also, using the full paint_bounds is allocating a larger texture than necessary because the child cannot draw outside the clip.  Instead, we should use the bounds of the rrect to for the size of the layer we save (skipping the margin we added to the paint bounds for the child).